crates/core/src/config.rs

```diff
@@ -42,62 +38,334 @@ const QUALIFIER: &str = "";
 const ORGANIZATION: &str = "The Freenet Project Inc";
 const APPLICATION: &str = "Freenet";
 
+#[derive(clap::Parser, Debug, Serialize, Deserialize)]
+pub struct ConfigArgs {
+    /// Node operation mode.
+    #[clap(value_enum, env = "MODE")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub mode: Option<OperationMode>,
+
+    /// Path to the configuration file.
+    #[clap(short, long)]
+    #[serde(skip)]
+    pub config: Option<PathBuf>,
+
+    #[clap(flatten)]
+    #[serde(flatten)]
+    pub gateway: GatewayArgs,
+    #[clap(value_parser, env = "TRANSPORT_KEYPAIR")]
+    pub transport_keypair: Option<PathBuf>,
+    #[serde(
+        with = "serde_option_log_level_filter",
+        skip_serializing_if = "Option::is_none"
+    )]
+    #[clap(long, env = "LOG_LEVEL")]
+    pub log_level: Option<tracing::log::LevelFilter>,
+    #[clap(flatten)]
+    #[serde(flatten)]
+    config_paths: ConfigPathsArgs,
+}
+
+impl Default for ConfigArgs {
+    fn default() -> Self {
+        Self {
+            mode: Some(OperationMode::Local),
+            config: None,
+            gateway: GatewayArgs {
+                address: Some(default_gateway_address()),
+                port: Some(default_gateway_port()),
+            },
+            transport_keypair: None,
+            log_level: Some(tracing::log::LevelFilter::Info),
+            config_paths: Default::default(),
+        }
+    }
+}
+
+impl ConfigArgs {
+    /// Parse the command line arguments and return the configuration.
+    pub fn build(mut self) -> std::io::Result<Config> {
+        let cfg = if let Some(path) = self.config.as_ref() {
```

Comment:
we should try to find a config in the default application config directory called `freenet.toml`  (see config paths for default application config, varies by platform) If not specified.

if it doesn't exist, we should save the config after loading for the first time so we can reload later